specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1096 @@
+If the message's destination is not the current subnet, the destination `Subnet` is checked to determine if the message needs to be propagated as a top-down or a bottom-up message, conveniently adding it in `TopDownMsgs`of the next subnet in the messages path to the destination, or by including it in the next checkpoint, respectively.
+- Increment the corresponding `AppliedNonces`.
+
+### Top-down messages
+When a new top-down cross-net message is sent to the `SCA` by calling `Fund()`, or `SendCross()` with a destination which is down in the hierarchy, the `SCA` of the source subnet:
+- Checks which child is the next subnet in the path to the messages destination.
+- Assigns to the message the subsequent `Nonce` to the latest one used for that subnet. Every time a new top-down message to the subnet arrives, the `SCA` increments a nonce that is unique for every cross-net message to that destination. These nonces determine the total order of arrival of cross-msgs to the subnet; without them, different consensus nodes could execute different orderings, leading to nondeterminism. 
+- Checks that that the `From` and `To` of the message include `HA` addresses with the right subnets. 
+- Stores the message to notify its propagation in the `TopDownMsgs` of the corresponding child subnet.
+- Locks in the `SCA` the number of native tokens included in the `value` of the message, increasing by the same amount the `CircSupply` of the subnet. These funds will be frozen until a bottom-up transaction releases them back to the parent. In this way, the `SCA` keeps track of the circulating supply of child subnets and is responsible for enforcing the firewall requirement in subnets (whenever native tokens want to be released from it, see [bottom-up messages](#bottom-up-messages)).
+
+When a new cross-net message to a subnet is included in its `TopDownMsgs`, the validators of the subnet are notified through their `CrossMsgPool` about this new top-down message. The `CrossMsgPool` waits for a `finalityThreshold` before proposing the message to be sure that the top-down message commitment can be considered final (this threshold may change according to the requirements of the subnet, and the specific consensus algorithm used by the parent. The threshold for BFT-like consensus may be `1` or close to `1`, while sync consensus algorithms like PoW may require larger values). After the `finalityThreshold`, validators in the subnet will check the latest `AppliedTopDownNonce` to fetch all unverified cross-net messages up to the latest one included in `TopDownMsgs`, and propose them for their inclusion in a block and subsequent execution.
```

Comment:
```suggestion
When a new cross-net message to a subnet is included in the `TopDownMsgs` in the parent subnet, the validators of the child subnet, who follow the parent consensus, are notified through their `CrossMsgPool` about this new top-down message. The `CrossMsgPool` waits for a `finalityThreshold` before proposing the message to be sure that the top-down message commitment can be considered final (this threshold may change according to the requirements of the subnet, and the specific consensus algorithm used by the parent. The threshold for BFT-like consensus may be `1` or close to `1`, while sync consensus algorithms like PoW may require larger values). After the `finalityThreshold`, validators in the child subnet will check the latest `AppliedTopDownNonce` to fetch all unverified cross-net messages up to the latest one included in `TopDownMsgs`, and propose them for their inclusion in a block and subsequent execution.
```

I may be getting this wrong, sorry. The `TopDownMsgs` is something a child operator sees in the ledger of the parent chain, right? They wait for it to be final, then they pluck it out and relay it to their child chain, unless `AppliedTopDownNonce` indicates this has already happened.

Or do these top down messages sit in the child ledger, unexecuted until finality is reached in the parent?

I'm also unclear what kind of proof is needed for them to be included and executed: are validators in the child subnet rejecting top down messages included in blocks by adversarial peers if they can't see them being finalized in the parent chain yet?

Reply:
The former. `AppliedTopDownNonce` aggregates all top-down messages directed to a child subnet. The `CrossMsgPool` in the subnet waits for finality before proposing it in a block of the child consensus. Full-nodes receiving blocks in the child subnet including these cross-messages check in the parent ledger if the cross-net message is final and part of `AppliedTopDownNonce` and won't accept blocks that don't satisfy the requirement.

In the next iteration we should consider including more succinct proofs to prevent the parent synchronization from delaying too much validators in the subnet.